src/adapters/adbutler.js

```diff
@@ -0,0 +1,146 @@
+    
+    var bids = params.bids || [],
+        callbackData = {},
+        zoneCount = {},
+        pageID = Math.floor(Math.random() * 10e6);
+
+    //Build and send bid requests
+    for (var i = 0; i < bids.length; i++) {
+      var bid = bids[i],
+        zoneID = utils.getBidIdParamater('zoneID', bid.params),
+        callbackID;  
+        
+      if(!(zoneID in zoneCount)){
+        zoneCount[zoneID] = 0;
+      }
+      
+      //build callbackID to get placementCode later  
+      callbackID = zoneID + '_' +zoneCount[zoneID];
+
+      callbackData[callbackID] = {};
+
+      callbackData[callbackID].placementCode = bid.placementCode;
+      callbackData[callbackID].sizes = bid.sizes;
+      callbackData[callbackID].maxCPM = utils.getBidIdParamater('maxCPM', bid.params);
+      callbackData[callbackID].minCPM = utils.getBidIdParamater('minCPM', bid.params);
+      
+      var adRequest = buildRequest(bid,zoneCount[zoneID],pageID);
+      zoneCount[zoneID]++;
+
+      adloader.loadScript(adRequest);
+    }
+      
+    //Define callback function for bid responses
+    $$PREBID_GLOBAL$$.adbutlerCB = function(aBResponseObject){
+        
+      var bidResponse = {},
+        callbackID = aBResponseObject.zone_id+'_'+aBResponseObject.place,
+        width = parseInt(aBResponseObject.width),
+        height = parseInt(aBResponseObject.height),
+        isCorrectSize = false,
+        isCorrectCPM = true,
+        CPM,minCPM,maxCPM;
+        
+      //Ensure that response ad matches one of the placement sizes.  
+      utils._each(callbackData[callbackID].sizes,function(size){
+        if(width === size[0] && height === size[1]){
+          isCorrectSize = true;
+        }
+      });
+
+      if (aBResponseObject.status === 'SUCCESS' && isCorrectSize) {
+        
+        CPM = aBResponseObject.cpm;
+        minCPM = callbackData[callbackID].minCPM;
+        maxCPM = callbackData[callbackID].maxCPM;
+        
+        //Ensure response CPM is within the given bounds
+        if(minCPM !== '' && CPM < parseFloat(minCPM)){
+          isCorrectCPM = false;
+        }    
+        if(maxCPM !== '' && CPM > parseFloat(maxCPM)){
+          isCorrectCPM = false;
+        }  
+        
+        if(isCorrectCPM){
+            
+          bidResponse = bidfactory.createBid(1);
+          bidResponse.bidderCode = 'adbutler';
+          bidResponse.cpm = CPM;
+          bidResponse.width = width;
+          bidResponse.height = height;
+          bidResponse.ad = aBResponseObject.ad_code;
+          bidResponse.ad += addTrackingPixels(aBResponseObject.tracking_pixels); 
+            
+        } else {
+          bidResponse = bidfactory.createBid(2);
+          bidResponse.bidderCode = 'adbutler';
+        } 
+          
+      } else {
+        bidResponse = bidfactory.createBid(2);
+        bidResponse.bidderCode = 'adbutler';
+      }
+        
+      bidmanager.addBidResponse(callbackData[callbackID].placementCode, bidResponse);
```

Comment:
Accessing `placementCode` when `callbackData[callbackID]` is `undefined` could also cause `TypeError`s. These particularly pop up when Karma is left in debug mode, which you can check with `gulp serve --watch` and then opening the developer console